stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -288,6 +388,150 @@ impl NakamotoBlock {
+    ///  block itself as processed.
+    pub fn set_block_processed(
+        staging_db_tx: &rusqlite::Transaction,
+        block: &StacksBlockId,
+    ) -> Result<(), ChainstateError> {
+        let update_dependents = "UPDATE nakamoto_staging_blocks SET stacks_attachable = 1
+                                 WHERE parent_block_id = ?";
+        staging_db_tx.execute(&update_dependents, &[&block])?;
+
+        let clear_staged_block =
+            "UPDATE nakamoto_staging_blocks SET processed = 1, processed_time = ?2
+                                  WHERE index_block_hash = ?1";
+        staging_db_tx.execute(
+            &clear_staged_block,
+            params![&block, &u64_to_sql(get_epoch_time_secs())?],
+        )?;
+
+        Ok(())
+    }
+
+    /// Notify the staging database that a given burn block has been processed.
+    /// This is required for staged blocks to be eligible for processing.
+    pub fn set_burn_block_processed(
+        staging_db_tx: &rusqlite::Transaction,
+        block: &BurnchainHeaderHash,
+    ) -> Result<(), ChainstateError> {
+        let update_dependents = "UPDATE nakamoto_staging_blocks SET burn_attachable = 1
+                                 WHERE burn_view = ?";
+        staging_db_tx.execute(&update_dependents, &[&block])?;
+
+        Ok(())
+    }
+
+    pub fn next_ready_block(
+        staging_db_conn: &Connection,
+    ) -> Result<Option<NakamotoBlock>, ChainstateError> {
+        let query = "SELECT data FROM nakamoto_staging_blocks
+                     WHERE burn_attachable = 1
+                       AND stacks_attachable = 1
+                       AND orphaned = 0
+                       AND processed = 0
+                     ORDER BY height ASC";
+        staging_db_conn
+            .query_row_and_then(query, NO_PARAMS, |row| {
+                let data: Vec<u8> = row.get("data")?;
+                let block = NakamotoBlock::consensus_deserialize(&mut data.as_slice())?;
+                Ok(Some(block))
+            })
+            .or_else(|e| {
+                if let ChainstateError::DBError(DBError::SqliteError(
+                    rusqlite::Error::QueryReturnedNoRows,
+                )) = e
+                {
+                    Ok(None)
+                } else {
+                    Err(e)
+                }
+            })
+    }
+
+    pub fn accept_block(
+        block: NakamotoBlock,
+        sortdb: &SortitionHandleConn,
+        staging_db_tx: &rusqlite::Transaction,
+    ) -> Result<(), ChainstateError> {
+        let recovered_miner_pk = block.header.recover_miner_pk().ok_or_else(|| {
+            warn!(
+                "Stacks block downloaded with unrecoverable miner public key";
+                "block_hash" => %block.header.block_hash(),
+            );
+            return ChainstateError::InvalidStacksBlock("Unrecoverable miner public key".into());
+        })?;
+
+        if sortdb
```

Comment:
I think this whole `if`-block needs to be dropped.  A node can accept any block back to the last PoX anchor block, especially while booting up.

Reply:
I think there needs to be two different acceptance paths for blocks:

1. Any block that has been committed to by a leader block commit -- leader block commits still commit to block hashes, but those are not the chain tip. These are used to accept blocks while booting up, catching up to the chain tip.
2. The if block in the code above: while operating at the chain tip, the only blocks that should be accepted are ones in recent tenures.